tests/functional/data_type/base_data_type_macro.py

```diff
@@ -0,0 +1,48 @@
+import os
+import pytest
+from dbt.tests.util import run_dbt, check_relations_equal, get_relation_columns
+
+class BaseDataTypeMacro:
+    @pytest.fixture(scope="class")
+    def packages(self):
+        return {"packages": [{"local": os.getcwd()}]}
+
+    def is_legacy(self):
+        return False
+
+    def test_check_types_assert_match(self, project):
+        run_dbt(['deps'])
+        run_dbt(['build'])
+        
+        # check contents equal
+        check_relations_equal(project.adapter, ["expected", "actual"])
+        
+        # check types equal
+        expected_cols = get_relation_columns(project.adapter, "expected")
+        actual_cols = get_relation_columns(project.adapter, "actual")
+        print(f"Expected: {expected_cols}")
+        print(f"Actual: {actual_cols}")
+        
+        
+        if not self.is_legacy():
+            assert expected_cols == actual_cols, f"Type difference detected: {expected_cols} vs. {actual_cols}"
+        # we need to be a little more lenient when mapping between 'legacy' and 'new' types that are equivalent
+        # e.g. 'character varying' and 'text'
+        elif expected_cols == actual_cols:
+            # cool, no need for jank
+            pass
+        else:
+            # this is pretty janky
+            for i in range(0, len(expected_cols)):
+                expected = project.adapter.Column(*expected_cols[i])
+                actual = project.adapter.Column(*actual_cols[i])
+                print(f"Subtle type difference detected: {expected.data_type} vs. {actual.data_type}")
+                if any((
+                    expected.is_string() and actual.is_string(),
+                    expected.is_float() and actual.is_float(),
+                    expected.is_integer() and actual.is_integer(),
+                    expected.is_numeric() and actual.is_numeric(),
+                )):
```

Comment:
Spit-balling an untested(!) idea:
Define the [rich comparison method](https://portingguide.readthedocs.io/en/latest/comparisons.html#rich-comparisons) `__eq__` within [columns.py](https://github.com/dbt-labs/dbt-core/blob/a642b20abc3b729cb04fd0616db065c5d0c19868/core/dbt/adapters/base/column.py) as:

```python
    def __eq__(self, other):
        return any((
                    self.is_string() and other.is_string(),
                    self.is_float() and other.is_float(),
                    self.is_integer() and other.is_integer(),
                    self.is_numeric() and other.is_numeric(),
                ))
```

It would take some thought and care to go down this route since this would literally override the behavior of `==` for Columns. You'd probably expand the implementation above to handle all the desired cases (like comparing additional properties too). Or define a ColumnDataType class and override the `__eq__` method for that thing instead.

Reply:
It's a neat idea. I'm disinclined to change `dbt-core` functionality to accomplish this, though. The goal of this "more lenient" type comparison is to acknowledge that there are some subtle difference between the previous value yielded by the `dbt_utils.type_{X}` macro, and the new ones I'm proposing in this PR. This logic says, as long as they're close enough, this _probably_ isn't a breaking change to `dbt-utils` functionality, and we should feel comfortable making the migration.

It is a good callout for me to try refactoring this logic, however.